crates/common/src/backup/rpc.rs

```diff
@@ -0,0 +1,152 @@
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use jsonrpsee::core::RpcResult;
+use jsonrpsee::proc_macros::rpc;
+use jsonrpsee::types::error::{INTERNAL_ERROR_CODE, INTERNAL_ERROR_MSG};
+use jsonrpsee::types::ErrorObjectOwned;
+use serde::{Deserialize, Serialize};
+use sov_db::ledger_db::{LedgerDB, SharedLedgerOps};
+
+use super::{BackupManager, CreateBackupInfo};
+
+/// Response from backup validation request
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ValidationResponse {
+    /// Path that was validated
+    pub backup_path: PathBuf,
+    /// Whether the backup at the path is valid
+    pub is_valid: bool,
+    /// Error message if validation failed
+    pub message: Option<String>,
+}
+
+/// Information about a specific backup
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct BackupInfoResponse {
+    /// Timestamp of the backup
+    pub timestamp: i64,
+    /// ID of the backup
+    pub backup_id: u32,
+    /// Size of the backup
+    pub size: u64,
+    /// Number of files related to the backup
+    pub num_files: u32,
+}
+
+#[rpc(client, server, namespace = "backup")]
+pub trait BackupRpc {
+    #[method(name = "create")]
+    async fn backup_create(&self, path: Option<PathBuf>) -> RpcResult<CreateBackupInfo>;
+
+    #[method(name = "validate")]
+    async fn backup_validate(&self, path: PathBuf) -> RpcResult<ValidationResponse>;
+
+    #[method(name = "info")]
+    async fn backup_info(
+        &self,
+        path: PathBuf,
+    ) -> RpcResult<HashMap<String, Vec<BackupInfoResponse>>>;
+}
+
+pub struct BackupRpcServerImpl {
+    backup_manager: Arc<BackupManager>,
+    ledger_db: LedgerDB,
+}
+
+impl BackupRpcServerImpl {
+    pub fn new(backup_manager: Arc<BackupManager>, ledger_db: LedgerDB) -> Self {
+        Self {
+            backup_manager,
+            ledger_db,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl BackupRpcServer for BackupRpcServerImpl {
+    async fn backup_create(&self, path: Option<PathBuf>) -> RpcResult<CreateBackupInfo> {
+        let l2_height = self
+            .ledger_db
+            .get_head_soft_confirmation_height()
+            .map_err(|e| {
+                ErrorObjectOwned::owned(
+                    INTERNAL_ERROR_CODE,
+                    INTERNAL_ERROR_MSG,
+                    Some(format!("{e}")),
+                )
+            })?
+            .unwrap_or_default();
+
+        self.backup_manager
+            .create_backup(path, l2_height)
```

Comment:
This is basically a DOS attack vector. Since creating a backup blocks block production, an attacker could choose to make frequent requests to create backups holding the locks for block production and at the same time causing the node storage to fill quickly. What do we do to mitigate this?

If devops can set this up in a way where the RPC is only callable from a specific subnet and make sure this RPC is not publicly available, then we're good.

Reply:
cc: @eyusufatik @Chinacolt